### PR TITLE
rpl: remove duplicate code by merging rpl_storing.c and rpl_nonstoring.c

### PR DESCRIPTION
**Pull request by cgundogan**
Looking at the files **rpl_storing.c** and **rpl_nonstoring.c** reveales, that there is a great amount of code duplication. Thus, any change to the current rpl implementation dictates to touch at least two files.

In this PR I merged the files **rpl_storing.c** and **rpl_nonstoring.c** into **rpl_control_messages.c** by *ifdef*ing the varying parts, because there are not much of them.
Furthermore, I removed the files **rpl_storing.h** and **rpl_nonstoring.h** as they were just introducing another layer in between.

As a positive side effect this PR also cuts a little bit in rom size.

### CODE REVIEW

**Comment by cgundogan** [comment]
@fabianbrandt what is your opinion on this?

**Comment by cgundogan** [comment]
@fabianbrandt what is your opinion on this?

**Comment by fabianbrandt** [comment]
Since the change is quiet big, Cenk and I discussed the PR with all changes involved in person. Actually it looks very promising. The initial idea behind splitting storing/non-storing mode was that both modes act very differently. However at the current state it seems, that this is simply not the case :) I will look more into detail in the code (hopefully soon), but after a quick overview this PR should save a lot of logic and LOCs.

One or two things to consider (since features are changed/erased by this PR):
- leaf/router mode as variable in dodag-structure (for RFC compliance)
- (for future) integration of P2P-mode in this new structure
- should we keep with the `#ifdef`-structure?

**Comment by fabianbrandt** [comment]
After looking more into the structure of this PR - it looks very promising and should help Cenk with future updates (+ see post above). If there are no objections, I would like to merge this :) @OlegHahm

File: sys/net/routing/rpl/rpl_control_messages.c [hunk 1]
```diff
@@ -10,16 +10,19 @@
  * @ingroup     rpl
  * @{
  *
- * @file        rpl_storing.c
- * @brief       RPL storing-mode
+ * @file        rpl_control_messages.c
```
Comment (OlegHahm) [inline sys/net/routing/rpl/rpl_control_messages.c]: Just remove this parameter while you're at it.

**Comment by cgundogan** [comment]
addressed comment by @OlegHahm and rebased to master

**Comment by cgundogan** [comment]
@OlegHahm I know it is a little bit time consuming as this PR is quite big, but I really would like your opinion on this, before the new network stack gets activated. Otherwise, I (or anyone else) would have to touch again 2 or more files in RPL to adjust the API or do other sorts of things.

**Comment by cgundogan** [comment]
It might help to throw a `vimdiff` on *rpl_storing.c* and *rpl_nonstoring.c*. After removing some DEBUGFs that are varying in text and moving around very little code, it gets quite obvious that both files have very much in common.

**Comment by LudwigOrtmann** [comment]
@cgundogan Oleg will be back only on Thursday.

**Comment by cgundogan** [comment]
@OlegHahm ping?

**Comment by cgundogan** [comment]
@BytesGalore can you also have a look at this PR and verify that this is running on your current hardware?

**Comment by BytesGalore** [comment]
@cgundogan successfully tested with storing and non storing MOP on 2 `iot-lab_M3` nodes root<->node.
Will have a look on the code now

File: sys/net/include/rpl.h [hunk 2]
```diff
@@ -109,32 +113,29 @@ void rpl_send_DAO(ipv6_addr_t *destination, uint8_t lifetime, bool default_lifet
...
- * differently in different modes, this function just sets the mutex and call the DIS
- * sending function of the chosen mode.
+ * This function sends a DIS message to a given destination or multicast-address.
  *
  * @param[in] destination       IPv6-address of the destination of the DIS. Should be a direct neighbor or multicast-address.
  *
  */
 void rpl_send_DIS(ipv6_addr_t *destination);
 
+#if RPL_DEFAULT_MOP != RPL_NON_STORING_MODE
```
Comment (BytesGalore) [inline sys/net/include/rpl.h]: please change to:
`RPL_DEFAULT_MOP != RPL_MOP_NON_STORING_MODE`

**Comment by BytesGalore** [comment]
maybe you could astyle the code

File: sys/net/routing/rpl/rpl_control_messages.c [hunk 3]
```diff
@@ -412,8 +465,9 @@ void rpl_send_DAO_ACK_mode(ipv6_addr_t *destination)
     uint16_t plen = ICMPV6_HDR_LEN + DAO_ACK_LEN;
     rpl_send(destination, (uint8_t *)icmp_send_buf, plen, IPV6_PROTO_NUM_ICMPV6);
 }
+#endif
 
-void rpl_recv_DIO_mode(void)
+void rpl_recv_DIO(void)
 {
```
Comment (BytesGalore) [inline sys/net/routing/rpl/rpl_control_messages.c]: Do we need to process the DIO when we are root?

**Comment by cgundogan** [comment]
addressed comment by @BytesGalore 

@BytesGalore I am not 100% sure if a root can omit an incoming DIO in our implementation. Further in the code (l:656) the root makes a decision to increment its trickle timer depending on the DIO

**Comment by BytesGalore** [comment]
@cgundogan I somehow missed that line (:eyeglasses:). 
As this seems to be the only root specific action performed on receiving a DIO, I propose to move it to the very beginning and omit the rest if we are root, I think the root doesn't need to join his own DODAG.

Also probably we should not update the DODAG version [here](https://github.com/cgundogan/RIOT/blob/rpl_refactor_modes/sys/net/routing/rpl/rpl_control_messages.c#L626) just because someone disseminated a higher DODAG version number in a DIO.
This should never happen, as the root is only allowed to do [it](https://tools.ietf.org/html/rfc6550#page-84).

But anyways my above comments are not really related to this PR. 
(maybe we can tackle them in followup PRs)

I like this PR and beside astyle and squashing the PR is reasonable and working, so I would ACK it.

**Comment by OlegHahm** [comment]
Can you explain/recall why `rpl_send_mutex` isn't required any more and how this relates to this PR?

**Comment by cgundogan** [comment]
@OlegHahm RPL runs in one thread so there is no synchronization needed anymore (and I doubt it was ever needed). See this PR #2191 . During refactoring I removed one layer of function calls, which rendered itself obsolete once storing and non-storing mode were merged into one file.
Currently we have a call stack similar to:
```c
void rpl_send_DIO(...) {
    mutex_lock();
    rpl_send_DIO_mode(...);
    mutex_unlock();
}
```
and `rpl_send_DIO_mode` was, depending on the MOP, either called in *rpl_storing.c* or *rpl_nonstoring.c*. I removed this extra layer and thus the mutexes went away.
This would close #2191

**Comment by cgundogan** [comment]
@BytesGalore, used astyle. If everything is still okay I will squash.

**Comment by BytesGalore** [comment]
@cgundogan nice :+1: from my point of view its ready
@OlegHahm any objections?

**Comment by OlegHahm** [comment]
Needs rebasing now.

**Comment by OlegHahm** [comment]
ACK after rebasing.

**Comment by OlegHahm** [comment]
btw I like: +196 −1,373

**Comment by cgundogan** [comment]
rebased and squashed

File: sys/net/include/rpl.h [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2013, 2014  INRIA.
+ * Copyright (C) 2013 - 2015  INRIA.
```
Comment (LudwigOrtmann) [inline sys/net/include/rpl.h]: Do you want to give your copyright away?

Comment (cgundogan) [inline sys/net/include/rpl.h]: This PR is just a major refactoring, there were no additional functionalities added by my side. Is it right to add myself to the Copyright section?

Comment (LudwigOrtmann) [inline sys/net/include/rpl.h]: You already imply the copyright worthiness by prolonging INRIA's.

Comment (LudwigOrtmann) [inline sys/net/include/rpl.h]: (In my personal opinion, refactoring are actual changes to the code and should be subject to copyright law. But as IANAL, do not take my word for it.)

Comment (cgundogan) [inline sys/net/include/rpl.h]: I added myself to the copyright then.

**Comment by cgundogan** [comment]
@OlegHahm @BytesGalore I rebased to master and squashed the commits. Are your ACKs still holding?

**Comment by OlegHahm** [comment]
Yes. And go!

**Pull request merged**